pkg/await/apps_statefulset_test.go

```diff
@@ -0,0 +1,1252 @@
+		expectedError error
+	}{
+		{
+			description: "[Revision 1] Should succeed after creating StatefulSet",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully creates and initializes StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetAdded(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetCreating(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetProgressing(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetReady(inputNamespace, inputName, targetService))
+
+				// Timeout. Success.
+				timeout <- time.Now()
+			},
+		},
+		{
+			description: "[Revision 2] Should succeed after updating StatefulSet",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully updates StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdate(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdating(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdatingWithActiveReplica(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdateSuccess(inputNamespace, inputName, targetService))
+
+				// Timeout. Success.
+				timeout <- time.Now()
+			},
+		},
+		{
+			description: "[Revision 1] Should fail if timeout occurs before successful creation",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully creates StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetAdded(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetCreating(inputNamespace, inputName, targetService))
+
+				// Timeout. Failure.
+				timeout <- time.Now()
+			},
+			expectedError: &timeoutError{
+				object: statefulsetCreating(inputNamespace, inputName, targetService),
+				subErrors: []string{
+					"Failed to observe the expected number of ready replicas",
+				}},
+		},
+		{
+			description: "[Revision 2] Should fail if timeout occurs before successful update rollout",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully updates StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdate(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdating(inputNamespace, inputName, targetService))
+
+				// Timeout. Failure.
+				timeout <- time.Now()
+			},
+			expectedError: &timeoutError{
+				object: statefulsetUpdating(inputNamespace, inputName, targetService),
+				subErrors: []string{
+					"Failed to observe the expected number of ready replicas",
+					".status.currentRevision does not match .status.updateRevision",
```

Comment:
We kind of got the wording wrong IMHO with `Deployment` -- worth wondering whether we can state this in a way that is closer to what people who don't know about the internals expect:

* Instead of the first, how about something like: `x out of y replicas succeeded readiness checks`
* Instead of the second  `StatefulSet controller failed to advance from revision %q to revision %q`

Reply:
I'm going to open an issue to follow on with this work. It's a good suggestion, but a little more than I want to change right now.